### ISSUE TITLE
Support JPA 2.1 stored procedures returning result sets [DATAJPA-1092]

### ISSUE THREAD

**Issue by spring-projects-issues** [spring-projects/spring-data-jpa#1433]
**[Dmitriy Korobskiy](https://jira.spring.io/secure/ViewProfile.jspa?name=dkroot)** opened **[DATAJPA-1092](https://jira.spring.io/browse/DATAJPA-1092?redirect=false)** and commented

Some databases: notably MS SQL and MySQL (but not Oracle) support SPs returning result sets (one or multiple) via simple SELECTs. In MS Transact-SQL this seems to be a pervasive pattern. JPA 2.1 spec explicitly supports mapping of SP result set(s) e.g. in 3.10.17.1 Named Stored Procedure Queries:

> A stored procedure may return more than one result set. As with native queries, the mapping of result sets can be specified either in terms of a resultClasses or as a resultSetMappings annotation element. If there are multiple result sets, it is assumed that they will be mapped using the same mechanism — e.g., all via a set of result class mappings or all via a set of result set mappings. The order of the specification of these mappings must be the same as the order in which the result sets will be returned by the stored procedure invocation. If the stored procedure returns one or more result sets and no resultClasses or resultSetMappings element has been specified, any result set will be returned as a list of type Object[]. The combining of different strategies for the mapping of stored procedure result sets is undefined.

This feature does not seem to be supported by Spring Data JPA: e.g. see https://stackoverflow.com/questions/31097667/illegalargumentexception-type-cannot-be-null. I spent some time trying to make it work with both 1.5.2 and the current 2.0 snapshot to no avail.

Here is a test example working via pure JPA 2.1 (executed in Spring Boot 1.5.2 app):

```
CREATE PROCEDURE tmp_demo (@arg INT)
AS
BEGIN
  SELECT @arg + 1 AS simple_result;
END;
```

CustomDemoRepository:

```
import DemoResult;
import java.util.List;
import javax.persistence.EntityManager;
import javax.persistence.ParameterMode;
import javax.persistence.StoredProcedureQuery;
import lombok.Data;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

@Service
@Data
@RequiredArgsConstructor(onConstructor = @__(@Autowired))
public class CustomDemoRepository {

  //region Injected beans (via a RequiredArgsConstructor)
  private final EntityManager em;
  //endregion

  public List<DemoResult> execStoredProdDirectly(Integer arg) {
    StoredProcedureQuery spq = em.createStoredProcedureQuery("tmp_demo", DemoResult.class);
    spq.registerStoredProcedureParameter("arg", Integer.class, ParameterMode.IN);
    spq.setParameter("arg", 42);
    return spq.getResultList();
  }
}
```

DemoResult:

```
import java.io.Serializable;
import javax.persistence.Entity;
import javax.persistence.Id;
import lombok.Data;
import lombok.NoArgsConstructor;

@Entity
@Data // = @ToString @EqualsAndHashCode @Getter-s @Setter-s (non-final) @RequiredArgsConstructor
@NoArgsConstructor
public class DemoResult implements Serializable {

  @Id private Integer simpleResult;
}
```

My thoughts:

1. I understand that Spring Data JPA tries to map an SP OUT parameter to a repository `@Procedure` method result. This is fine, but when such a method is expected to return a list (List\<DomainObject>) it means that developers need to get a result set out of an SP. In MS SQL result sets can be returned implicitly without declaring any additional SP params.
2. Multiple result sets are used more rarely and support for those could be done separately from a single result set.

---

**Affects:** 1.11.1 (Ingalls SR1)

**Reference URL:** https://stackoverflow.com/questions/31097667/illegalargumentexception-type-cannot-be-null

**Issue Links:**
- [DATAJPA-1555](https://jira.spring.io/browse/DATAJPA-1555) InvalidDataAccessApiUsageException occurs when using `@Procedure` annotation
 (_**"is duplicated by"**_)

2 votes, 5 watchers

**Comment by netdragonboberb** [comment]
For the thousands of others reading this that are up against the same brick wall, to another library: I had the most luck using a straight Hibernate Session access with createQuery/createNativeQuery and setParameterList, however due to different SQL language types PLPGSQL vs HQL, it's still tricky and you'll get things like mismatch of array notation you have to work through (e.g. missing brackets for postgres on createNativeQuery)

**Comment by netdragonboberb** [comment]
Follow-up: I thought I should say what I ended up with. It turned out I didn't need setParameterList, but some of you may...
Since this is for postgres, you may be using table type parameters in other sql implementations (e.g. mysql or mssql) and have to change this a bit.

I believe you'll need it for an IN clause (correct me if I'm wrong)

e.g. if you have (warning: This may not be the exact right notation)

`
			var session = entityManager.unwrap( Session.class );
			var sql = "select * from foo where bar in :months)";
			var spq = session.createNativeQuery(sql, Foo.class);
			spq.setParameterList("months", dates.toArray(), DateArrayType.INSTANCE); // DateArrayType from types_52
			return spq.getResultList();`

However, because this is just an array parameter I was able to do this:

`			var session = entityManager.unwrap( Session.class );
			var sql = "select * from my_function(:months, :some_other_var)"; 
			var spq = session.createNativeQuery(sql, Foo.class);
			spq.setParameter("months", dates.toArray(), DateArrayType.INSTANCE); // DateArrayType from types_52
			spq.setParameter("some_other_var", 100);
			return spq.getResultList();`

effectively, it's doing something similar to:
`session.createNativeQuery("select * from my_function(array [cast('20180101' as date),cast('20180102' as date)],100)", Foo.class);`

Here is the postgres function (with changes to remove specific vars):


`
create or replace function my_function(months date[], some_other_var integer) /* pass first day of each month */ returns setof foo
as $$
begin
	return query
	with f_months as (select distinct d from unnest(months) as d)
	select
			f.*
	from foo f
	inner join  f_months cdate on
	(f.effective_from is null or date_trunc('month', f.effective_from) <= date_trunc('month', cdate.d) )
			and (f.effective_to is null or date_trunc('month', f.effective_to) > date_trunc('month', cdate.d) )
	where
		brg.some_other_var = some_other_var
		and f.is_removed = false
		and f.is_active = true;

end;
$$
language plpgsql;

-- Example:
-- SELECT public.my_function(array['2020-10-01', '2020-09-01']::date[],10100);

`



Warning: Due to changing the code a bit so it was more general, I can't guarantee it didn't break anything but if it did you shouldn't have to change much

**Comment by netdragonboberb** [comment]
If you absolutely run up against a wall, you can do something like this and then just toss it into the sql directly:

```
	private String getFormattedArray(Collection<Date> dates)
	{
		var strings = new ArrayList<String>();
		for (var date: dates)
		{
			strings.add("cast('" + date.toString() + "' as date )");
		}

		return "array [" + String.join(",", strings) + "]";
	}

```

**Comment by schauder** [comment]
Triage result: This would be a nice feature but won't be implemented in the near future. 

A workaround is available: Execute the query in a custom method using either an injected `EntityManager` or a JDBC `Connection` possibly wrapped in a `JdbcTemplate`.

**Comment by mp911de** [comment]
Implemented via #1959
